src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -551,6 +552,77 @@ _public_ int sd_pid_notify_with_fds(
         return r;
 }
 
+_public_ int sd_notify_barrier(int timeout) {
+        union sockaddr_union sockaddr;
+        char byte = 0;
+        struct iovec iovec = {
+                .iov_base = &byte,
+                .iov_len = 1,
+        };
+        struct msghdr msghdr = {
+                .msg_iov = &iovec,
+                .msg_iovlen = 1,
+                .msg_name = &sockaddr,
+        };
+
+        _cleanup_close_pair_ int pipe_fd[2] = { -1, -1 };
+        _cleanup_close_ int fd = -1;
+        struct cmsghdr *cmsg = NULL;
+        const char *e;
+        int r;
+
+        e = getenv("NOTIFY_SOCKET");
+        if (!e)
+                return 0;
+
+        r = sockaddr_un_set_path(&sockaddr.un, e);
+        if (r < 0)
+                goto finish;
+        msghdr.msg_namelen = r;
+
+        fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+        if (fd < 0) {
+                r = -errno;
+                goto finish;
+        }
+
+        if (pipe2(pipe_fd, 0) < 0) {
+                r = -errno;
+                goto finish;
+        }
+
+        msghdr.msg_controllen = CMSG_SPACE(sizeof(int));
+
+        msghdr.msg_control = alloca0(msghdr.msg_controllen);
+
+        cmsg = CMSG_FIRSTHDR(&msghdr);
+        cmsg->cmsg_level = SOL_SOCKET;
+        cmsg->cmsg_type = SCM_RIGHTS;
+        cmsg->cmsg_len = CMSG_LEN(sizeof(int));
+
+        memcpy(CMSG_DATA(cmsg), &pipe_fd[1], sizeof(int));
+
+        struct pollfd pfd = {
+                .fd = pipe_fd[0],
+                .events = POLLHUP,
```

Comment:
to avoid confusion, assign 0 here. poll() implicitly always watches for POLLHUP hence specifying it explicitly is a bit misleading as it suggests it would actually matter, but it does not... (a comment here saying so is of course welcome)